src/geometry/manifold/algebra/left_invariant_derivation.lean

```diff
@@ -0,0 +1,195 @@
+/-
+Copyright © 2020 Nicolò Cavalleri. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Nicolò Cavalleri
+-/
+
+import geometry.manifold.derivation_bundle
+
+/-!
+
+# Left invariant derivations
+
+In this file we define the concept of left invariant derivation for a Lie group. The concept is
+analogous to the more classical concept of left invariant vector fields, and it holds that the
+derivation associated to a vector field is left invariant iff the field is.
+
+Moreover we prove that `left_invariant_derivation I G` has the structure of a Lie algebra, hence
+implementing one of the possible definitions of the Lie algebra attached to a Lie group.
+
+-/
+
+noncomputable theory
+
+open_locale lie_group manifold derivation
+
+variables {𝕜 : Type*} [nondiscrete_normed_field 𝕜]
+{E : Type*} [normed_group E] [normed_space 𝕜 E]
+{H : Type*} [topological_space H] (I : model_with_corners 𝕜 E H)
+(G : Type*) [topological_space G] [charted_space H G] [monoid G] [has_smooth_mul I G] (g h : G)
+
+-- Generate trivial has_sizeof instance. It prevents weird type class inference timeout problems
+local attribute [nolint instance_priority, instance, priority 10000]
+private def disable_has_sizeof {α} : has_sizeof α := ⟨λ _, 0⟩
+
+/--
+Left-invariant global derivations.
+
+A global derivation is left-invariant if it is equal to its pullback along left multiplication by
+an arbitrary element of `G`.
+-/
+structure left_invariant_derivation extends derivation 𝕜 C^∞⟮I, G; 𝕜⟯ C^∞⟮I, G; 𝕜⟯ :=
```

Comment:
My final question about this definition: are you sure a `structure` is what we want here?

It means we need a lot of boiler plate and we also lose the information that what we actually have is a Lie subalgebra of all derivations.

What would happen if we tried:
```lean
def left_invariant_derivation : lie_subalgebra 𝕜 (derivation 𝕜 C^∞⟮I, G; 𝕜⟯ C^∞⟮I, G; 𝕜⟯) :=
{ carrier   := { D | ∀ g, 𝒅ₕ(smooth_left_mul_one I g) (derivation.eval_at 1 D) = derivation.eval_at g D},
  zero_mem' := sorry,
  add_mem'  := sorry,
  smul_mem' := sorry,
  lie_mem'  := sorry, }
```

Reply:
I believe this problem you propose holds for every algebraic structure, and I thought the current trend was reformulating statements about substructures in terms of morphisms, but I might have understood wrong. I mean the problem is analogous to the fact that bundled continuous functions do not allow to state that they are a subspace of all functions. This was not a problem because you can say there is a monomorphism into all functions. We could do as with continuous functions and have both bundled and unbundled if you think need to use subalgebras (but I think it is better not to). I strongly believe there should be an implementation as a structure, being it an implementation of the Lie algebra of a Lie group (at least in the real case).

Reply:
I think I would go for a `lie_subalgebra` but I see the case for a `structure` so let's leave it as it stands.

Reply:
Ok, btw the first time someone needs to talk about a subalgebra it is easy to refactor: one just needs to write the property as a definition `is_left_invariant_derivation`, and then you use that in the field of the structure `left_invariant_derivation`!